### PR TITLE
test: add Pester test infrastructure

### LINKED ISSUE #72: test: Pester テスト基盤を導入する
Blocked by #63

## Background

このリポジトリの検証手段は、Markdown / スペルの lint と PowerShell の静的解析
（PSScriptAnalyzer）だけである。静的解析は「書き方の問題」を見つけるが、
「入力に対して正しい出力を返すか」は検証できない。

検証手段を決めた時点では、それで十分だった。当時のリポジトリにあった
PowerShell はインストーラの呼び出しの羅列で、単体で検証できる純粋なロジックが
無かったためである。

その後の起票で前提が変わった。宣言的移行ロードマップは、入出力が明確な
ロジックを 2 件導入する。

- `scripts/Build-Configurations.ps1`: `dsc.yaml` を解析して `import.json` と
  未適用リソース一覧を生成する。ソース単位のグループ化、決定的な出力順、
  パッケージとそれ以外のリソースの振り分けなど、テストしやすく壊れやすい
  ロジックの塊になる
- `scripts/Test-PackageIds.ps1`: ID の存在確認を行い、「不在」と「判定不能」を
  区別して集計する。この区別こそが価値なので、区別が壊れたことを検出できる
  必要がある

また `libs/os-guard.ps1` の `Test-OsSupport` は、ビルド番号とエディションから
サポート状況を返す純粋な判定関数であり、既にテスト対象になりうる。

この issue はテスト基盤の導入のみを担当する。個々のロジックに対するテストは、
そのロジックを実装する issue が追加する。

## Proposed change

1. Pester によるテスト基盤を導入する。
   - テストの置き場所を `tests/powershell/` とする
   - Pester のバージョンを固定する方法を決める（`Install-Module` の
     `-RequiredVersion`、`PSDepend` 相当の仕組み、CI でのバージョン指定など）。
     選んだ方法と理由を記録すること
2. 少なくとも 1 件の実テストを含める。対象は実装時点で存在する純粋な
   ロジックとし、`libs/os-guard.ps1` の `Test-OsSupport` を第一候補とする。
   `Get-CimInstance` と `Get-ItemProperty` を Pester の `Mock` で差し替え、
   ビルド番号とエディションの組み合わせに対して期待する Tier が返ることを
   検証する。境界値（Windows 11 の下限 22000、Windows 10 22H2 の 19045、
   それ未満）を含めること。
3. CI にテストジョブを追加する。`pwsh` が必要なため `ubuntu-latest` を使う。
   PSScriptAnalyzer ジョブと同じ扱いにできるなら、ジョブを分けるか統合するかは
   実装者が判断してよい。
4. `pre-push-validate` と `post-fix-validate` にテスト実行を追加する。
   コマンド文字列を変更した場合は、`.github/idd/config.json` の `commands` と
   `.github/instructions/idd-overview-core.instructions.md` の Project commands
   表を同一 PR で更新し、両者を一致させること。
5. テストの書き方と実行方法を `docs/` に記載する。新しいスクリプトを追加する
   ときにテストも追加する、という方針を明記すること。

## Acceptance criteria

- `tests/powershell/` 配下に Pester のテストファイルが存在する
- Pester のバージョン固定方法が決まっており、選択理由が記録されている
- `Test-OsSupport` に対するテストが存在し、ビルド番号 22000 / 19045 /
  それ未満の各境界について期待する結果を検証している
- 当該テストが `Get-CimInstance` と `Get-ItemProperty` を Mock しており、
  実行するマシンの OS に依存しない
- ローカルでテストを実行して全件成功する
- CI にテストジョブが追加され、PR 上で green になる
- 意図的にテストを失敗させた状態で CI が赤くなることを確認している
  （確認結果を PR 説明に記載する）
- `pre-push-validate` と `post-fix-validate` にテスト実行が含まれている
- コマンド文字列を変更した場合、`.github/idd/config.json` の `commands` と
  Project commands 表が一致している
- テストの書き方と実行方法が `docs/` に記載されている
- PSScriptAnalyzer がテストファイルに対して指摘を出さない
- `pre-push-validate` コマンドがクリーンな作業ツリーに対して exit 0 で完了する

## IDD ブートストラップ

IDD 導入ロードマップ #55 が完了するまでは、実行フローを upstream 側から
読み込むこと。marker prefix は `setup-windows`。

```text
https://raw.githubusercontent.com/kurone-kito/idd-skill/main/idd-template/docs/idd-workflow.md
```

---

_Autopilot suitability: 4 / 5 -- higher is more autopilot-suitable;
below the configured floor is human-oriented._

_Effort: M -- author-estimated size; a soft autopilot
selection tie-breaker only._

### PR DESCRIPTION
**Pull request by kurone-kito**
## Summary

Closes #72.

This repository's only verification so far has been static (lint,
cspell, PSScriptAnalyzer) — none of it checks "does this return the
right output for this input." The declarative-migration roadmap
(#62) is about to add real input/output logic (`Build-Configurations.ps1`
in #66, `Test-PackageIds.ps1`), so this PR introduces Pester as the
unit-test layer before those land.

## What changed

- **`tests/powershell/os-guard.Tests.ps1`** (new): tests
  `libs/os-guard.ps1`'s `Test-OsSupport` across its boundary and gap
  values — build 22000 (Windows 11 floor, Pro/Home), build 19045
  (Windows 10 22H2 floor, Pro/Home), build 19044 (one below 22H2),
  build 21999 (the unnamed gap between 22H2 and Windows 11), build
  17763 (Windows Server 2019 floor), and build 10239 (pre-Windows-10).
  Data-driven via Pester's `-ForEach`.
- **`.github/workflows/lint.yml`**: new `pester` job (`ubuntu-latest`,
  `Install-Module -Name Pester -RequiredVersion 6.0.1 -Force -Scope
  CurrentUser -Repository PSGallery`, matching the existing
  `powershell-analyzer` job's version-pin pattern). Kept as its own
  job rather than folded into `powershell-analyzer` — different tools,
  different purpose, and `powershell-analyzer` isn't referenced
  anywhere else so renaming/merging it had no upside.
- **`.github/idd/config.json`** / **`idd-overview-core.instructions.md`**:
  `pre-push-validate` and `post-fix-validate` both gained
  `pwsh -c "Invoke-Pester -Path ./tests/powershell -CI"`. Verified
  programmatically (not by eye) that the two command strings stay
  character-identical after the edit.
- **`docs/testing.md`** (new): when to add a test, where tests live,
  the Pester version-pin rationale, how to run tests locally, and the
  `Get-CimInstance` mocking pattern below.
- **`.gitignore`**: `testResults.xml` (`Invoke-Pester -CI`'s
  `TestResult.Enabled` output).

## The `Get-CimInstance` mocking problem

`Get-CimInstance` ships in the Windows-only `CimCmdlets` module, so it
doesn't exist on `ubuntu-latest` (or in this Linux dev environment).
Pester's `Mock` requires the target command to exist before it can
mock it, so a naive `Mock Get-CimInstance { ... }` fails with
`CommandNotFoundException` — confirmed by a throwaway spike before
writing the real suite. The fix: declare a global placeholder function
before dot-sourcing the file under test:

```powershell
BeforeAll {
  function global:Get-CimInstance { }
  . (Join-Path -Path $PSScriptRoot -ChildPath '..' -AdditionalChildPath '..', 'libs', 'os-guard.ps1')
}
```

`Mock Get-CimInstance { ... }` then replaces this placeholder exactly
as it would replace the real cmdlet on Windows. Documented in both the
test file and `docs/testing.md` so the next test added to this suite
doesn't have to rediscover it. `Get-ItemProperty` needed no such
placeholder — it exists cross-platform even though the `HKLM:`
path it targets doesn't, so it mocks normally.

## The `$ErrorActionPreference` problem (found and fixed during this PR)

The first push of this branch (`5d93826`) went red on CI —
[run 30717940698](https://github.com/kurone-kito/setup.windows/actions/runs/30717940698/job/91416695397),
`pester` job — while passing locally. Root cause: GitHub Actions'
`shell: pwsh` steps implicitly set `$ErrorActionPreference = 'Stop'`
for the whole step. `Test-OsSupport`'s unsupported-OS branches signal
via a non-terminating `Write-Error`; under `Stop`, that becomes a
terminating exception instead, so the three unsupported-branch cases
(builds 19044, 21999, 10239) blew up mid-function instead of
returning. A plain local `pwsh -c` invocation defaults to `Continue`,
so it never surfaced there.

Reproduced locally by wrapping the same `Invoke-Pester` call in a
script with `$ErrorActionPreference = 'Stop'` set first — reproduced
the exact 3-case failure. Fixed in `3c672f7` by resetting the
preference to `'Continue'` in `BeforeAll`, confirmed the fix under the
same reproduction (8/8 pass), then confirmed on CI itself
(`3c672f7`'s push went green).

**Trade-off, disclosed rather than hidden**: `'Continue'` also
swallows any *unexpected* non-terminating error from code under test.
If `Test-OsSupport` ever started emitting `Write-Error` on a
*supported* branch, this suite wouldn't flag it as a new failure mode
— it would just see the branch's other assertions pass or fail on
their own merits. Accepted here because the alternative (asserting on
specific error records) would meaningfully complicate every future
test in this suite for a failure mode `Test-OsSupport`'s current
design doesn't create.

## Verified: CI actually goes red on a real failure

Two separate demonstrations, both on this branch before opening the
PR:

1. **Accidental** (the `$ErrorActionPreference` bug above): CI went
   red on `5d93826` from a genuine suite/environment mismatch, not a
   contrived one — arguably stronger evidence the suite catches real
   problems than a synthetic failure would be.
2. **Deliberate** (issue #72's own acceptance criterion): flipped the
   Windows-11-Pro case's `ExpectedTier` from `1` to `2` in `35233e6`,
   pushed, and confirmed
   [run 30718286562](https://github.com/kurone-kito/setup.windows/actions/runs/30718286562/job/91417610090)'s
   `pester` job failed (`Tests Passed: 7, Failed: 1`) while `lint` and
   `powershell-analyzer` stayed green in the same run. Reverted in
   `7b0c9c2`, confirmed CI green again
   ([run 30718349518](https://github.com/kurone-kito/setup.windows/actions/runs/30718349518)).

## Verification

- `markdownlint-cli2 "**/*.md"` — 0 issues
- `cspell lint "**" --no-progress` — 0 issues
- `Invoke-ScriptAnalyzer -Path . -Recurse -Settings
  ./PSScriptAnalyzerSettings.psd1 -EnableExit` — exit 0, including
  against the new test file (an early standalone PSA run against just
  `tests/powershell/os-guard.Tests.ps1` caught one
  `PSAvoidUsingPositionalParameters` info-level finding on `Join-Path`,
  fixed before it reached the full-suite run)
- `pwsh -c "Invoke-Pester -Path ./tests/powershell -CI"` — 8/8 pass
  locally, exit 0 (verified the actual process exit code directly,
  not through a pipeline where `$?` would reflect the wrong command)
- `.github/idd/config.json`'s and `idd-overview-core.instructions.md`'s
  `pre-push-validate`/`post-fix-validate` strings verified
  character-identical via a small Python diff, not by eye

## Test plan

- [x] `tests/powershell/` contains a Pester test file
- [x] Pester version-pin method decided (`-RequiredVersion 6.0.1`) and
      rationale recorded in `docs/testing.md`
- [x] `Test-OsSupport` has tests covering build 22000 / 19045 / below,
      plus the 21999 gap and 17763 Server floor
- [x] The test mocks `Get-CimInstance` and `Get-ItemProperty`, so it
      doesn't depend on the OS it runs on
- [x] All tests pass locally
- [x] CI test job added and green on this PR
- [x] Deliberately broke a test and confirmed CI goes red (see above,
      both the deliberate case and the accidental one that surfaced a
      real bug in this PR's own first draft)
- [x] `pre-push-validate` and `post-fix-validate` include test
      execution
- [x] `.github/idd/config.json` `commands` and the Project commands
      table match exactly (verified programmatically)
- [x] Test-writing and how-to-run docs added under `docs/`
      (`docs/testing.md`)
- [x] PSScriptAnalyzer reports nothing for the test file
- [x] `pre-push-validate` exits 0 on a clean worktree

## Summary by CodeRabbit

* **Tests**
  * Added automated PowerShell tests covering supported Windows client and server versions, unsupported versions, and boundary cases.
  * Validation now runs these tests alongside existing checks before pushes and after fixes.
  * Added continuous integration coverage for PowerShell tests.

* **Documentation**
  * Added guidance for running, maintaining, and troubleshooting PowerShell tests locally and in CI.

### CODE REVIEW

**Comment by kurone-kito** [comment]
_claude-code-01u4axbp: review triage snapshot — IDD automation marker. Do not edit._

**Comment by kurone-kito** [comment]
advisory-wait: claude-code-01u4axbp 7b0c9c2f65d50c9ac90acd1a7e621beb20a245bb 2026-08-01T21:06:56Z

File: tests/powershell/os-guard.Tests.ps1 [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+  # GitHub Actions' `shell: pwsh` steps implicitly set
+  # $ErrorActionPreference = 'Stop' for the whole step, which turns
+  # Test-OsSupport's non-terminating Write-Error calls (its documented
+  # signal for the unsupported-OS branches) into terminating exceptions
+  # here -- unlike a plain local `pwsh -c` invocation, where the
+  # session default of 'Continue' lets those branches return normally.
+  # Reset it explicitly so this suite behaves the same in both places.
+  $ErrorActionPreference = 'Continue'
+}
+
```
[the remark this answers is not in the document]
Comment (kurone-kito) [reply tests/powershell/os-guard.Tests.ps1:24]: Fixed in a91b1bb: added AfterAll to remove the global Get-CimInstance placeholder (via `Remove-Item -Path function:Get-CimInstance` -- the scope-qualified `function:global:Get-CimInstance` path silently no-ops, confirmed empirically) and restore $ErrorActionPreference to its pre-suite value. Verified with a stricter local check: started the session with $ErrorActionPreference = 'Stop', ran the suite, confirmed both are back to their original state afterward, and the suite itself still passes 8/8.

File: .github/workflows/lint.yml [hunk 2]
```diff
@@ -28,3 +28,13 @@ jobs:
       - name: Run PSScriptAnalyzer
         shell: pwsh
         run: Invoke-ScriptAnalyzer -Path . -Recurse -Settings ./PSScriptAnalyzerSettings.psd1 -EnableExit
+  pester:
+    runs-on: ubuntu-latest
+    steps:
+      - uses: actions/checkout@v7
```
[the remark this answers is not in the document]
Comment (kurone-kito) [reply .github/workflows/lint.yml:34]: Fixed in 135f243: added persist-credentials: false to the pester job's checkout step, matching the pattern already used by idd-advisory-convergence.yml and post-merge-cleanup.yml. The lint and powershell-analyzer jobs in this same file have the same gap but are pre-existing and out of scope for this PR.

**Comment by kurone-kito** [comment]
advisory-wait: claude-code-01u4axbp 135f243b4146d62b1b903c0f6de4bb78fff12a42 2026-08-01T21:17:06Z

**Comment by kurone-kito** [comment]
_claude-code-01u4axbp: review triage snapshot — IDD automation marker. Do not edit._

**Comment by kurone-kito** [comment]
**Rejected** — coderabbitai[bot] did not review HEAD 135f243b4146d62b1b903c0f6de4bb78fff12a42 (review limit reached / rate limited); this is not a completed review (source: #issuecomment-5153437760)

**Comment by kurone-kito** [comment]
_claude-code-01u4axbp: review triage snapshot — IDD automation marker. Do not edit._

**Comment by kurone-kito** [comment]
_claude-code-01u4axbp: review triage snapshot — IDD automation marker. Do not edit._
